src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1217,145 +900,215 @@ ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::limit_init_tm_tx(
+            break;
+        case tm_stm::not_leader:
+        case tm_stm::partition_not_found:
+            reply.ec = tx::errc::not_coordinator;
+            break;
+        case tm_stm::timeout:
+            reply.ec = tx::errc::timeout;
+            break;
+        }
+    }
+    co_return reply;
+}
 
 ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
   ss::shared_ptr<tm_stm> stm,
   model::term_id term,
   kafka::transactional_id tx_id,
   std::chrono::milliseconds transaction_timeout_ms,
   model::timeout_clock::duration timeout,
-  model::producer_identity expected_pid) {
-    auto r0 = co_await get_tx(term, stm, tx_id, timeout);
+  std::optional<model::producer_identity> expected_pid) {
+    auto tx_result = co_await find_and_try_progressing_transaction(
+      term, stm, tx_id, timeout);
 
-    if (!r0.has_value()) {
-        if (r0.error() != tx::errc::tx_not_found) {
+    if (tx_result.has_error()) {
+        if (tx_result.error() == tx::errc::tx_not_found) {
+            co_return co_await allocate_new_producer_id(
+              stm, term, std::move(tx_id), transaction_timeout_ms, timeout);
+        }
+        vlog(
+          txlog.warn,
+          "[tx_id={}] error getting transaction metadata: {}",
+          tx_id,
+          tx_result.error());
+        co_return init_tm_tx_reply{tx_result.error()};
+    }
+    // mapping for current transactional id already exists
+    auto tx = std::move(tx_result.value());
+
+    if (!is_valid_producer(tx, expected_pid)) {
+        vlog(
+          txlog.info,
+          "[tx_id={}] producer with expected pid: {} for {} is invalid",
+          tx_id,
+          expected_pid,
+          tx);
+        co_return init_tm_tx_reply{tx::errc::invalid_producer_epoch};
+    }
+
+    switch (tx.status) {
+    case ongoing: {
+        vlog(txlog.info, "[tx_id={}] tx is ongoing, aborting", tx_id);
+        auto abort_result = co_await do_abort_tm_tx(term, stm, tx, timeout);
+        if (!abort_result) {
             vlog(
               txlog.warn,
-              "[tx_id={}] error getting transaction: {}",
+              "[tx_id={}] error rolling previous transaction",
               tx_id,
-              r0.error());
-            co_return init_tm_tx_reply{tx::errc::not_coordinator};
+              tx.status,
+              abort_result.error());
+            co_return init_tm_tx_reply{abort_result.error()};
         }
+        co_return init_tm_tx_reply{tx::errc::concurrent_transactions};
+    }
+    case empty:
+    case tombstone:
+    case completed_commit:
+    case completed_abort: {
+        co_return co_await increase_producer_epoch(
+          tx.id,
+          tx.pid,
+          tx.last_pid,
+          expected_pid,
+          stm,
+          term,
+          transaction_timeout_ms,
+          timeout);
+    }
+    case preparing_abort:
+    case expired:
+    case preparing_commit:
+        // TODO: make sure we move forward with those bad boys
```

Comment:
did you mean to retain this comment :D

Reply:
Is this possible though? doesn't find_and_try_progressing_transaction() roll forward to completed_commit/abort?

Reply:
I _think_ concurrent_transactions semantics are different, it should be returned before rolling forward/back in the above section of code so client can retry while rollback/forward is in progress.

Reply:
agreed, that is indeed impossible in this case, i will add returning an error in this case. Maybe we can change how we handle progressing transaction in next PR. Change the `find_and_try_progressing_transaction` to try progressing in the background ?

Reply:
sure.